enhancements/kube-apiserver/stale-clusteroperators-detection.md

```diff
@@ -0,0 +1,87 @@
+---
+title: detection-and-reporting-of-stale-clusteroperators
+authors:
+  - "@mfojtik"
+reviewers:
+  - "@deads2k"
+approvers:
+  - "@deads2k"
+creation-date: 2022-01-31
+last-updated: 2022-01-31
+---
+
+# Detection and reporting of stale cluster operators
+
+## Summary
+
+In OpenShift 4, every active operator included in the core payload (cluster operator) updates its conditions in "
+clusteroperator" resource. Conditions determine whether specific core components are Degraded or Available or if they
+are making progress or if their state causes the cluster to be not upgradeable. These conditions are essential for
+cluster administrators to determine the health of their clusters, and they are usually the first thing Red Hat support
+is going to look at in case there is a problem with the cluster.
+
+Unfortunately, in some edge cases, these conditions might not reflect the current state of the cluster. For example,
+since they depend on the running operator, these conditions might become stale and outdated if the operator is not
+running or cannot contact the API server. And while the API server might still work, the result
+of `oc get clusteroperators` does not represent the cluster's current state.
+
+Therefore, we need a detection mechanism to help detect in-active operators or operators that are not actively updating
+the conditions.
+
+## Motivation
+
+In the past, there were multiple cases when either kubelet or CRI-O failed to run pods, either by KCM misfunctioning or
+some other system error. However, the static pods succeeded in running, making the Kubernetes API server available and
+accessible with `oc`. When support or engineers ran `oc get clusteroperators` against this cluster, they saw all
+operators reporting available, non-degraded and upgradeable status, but the truth was that none of the operators that
+provided these conditions were running.
+
+### Goals
+
+* Detect in-active cluster operators
+* Provide more accurate condition messages and status
+
+### Non-Goals
+
+* Make a new operator that detects other operators status
+* Create a heart-beat mechanism
+
+## Proposal
+
+To implement stale operator detection, we need a controller process that will run close to the Kubernetes API server,
+preferably using a local connection and client-based certificate for authentication to avoid dependencies on working
+auth and network when the cluster is unstable.
+
+The controller will watch the clusteroperator resource and check the `status.[]conditions.lastTransitionTime` field for
+Degraded, Available, Progressing, and Upgradeable conditions. Suppose the "lastTransitionTime" field is not updated for
+10 minutes. In that case, the controller will change the human-readable condition message
+to: `Operator checking for stale status, the active operator will reset this message: <original message>`.
```

Comment:
for a healthy operator that's not experiencing any changes, aren't we going to be asserting this message every 10 minutes and then waiting for it to be reset back?

I guess the reset back will be quick (the operator should immediately detect the change and reassert the value) but it's definitely new churn.

it also loses the ability to know when the operator *really* last changed a condition(e.g. when did it enter progressing?)

Reply:
@deads2k i realize... should operators change lastTransitionTime when they are not transitioning states, but only update message?

Reply:
covered above, lastTransitionTime should not change just because the message is updated.

i still have a small concern about churning the messages which may look strange in any sort of dashboard that's reporting the message strings.  It's not a showstopper, but it doesn't feel like an ideal UX for the end-user.

Reply:
@deads2k @mfojtik this thread never really got resolved, is this the behavior that's being baked into the proposed implementation?  flapping status messages?

Reply:
> @deads2k @mfojtik this thread never really got resolved, is this the behavior that's being baked into the proposed implementation? flapping status messages?

Yes, the message gets a prefix "Checking for stale status, the active operator will reset this message: " and the existing message.  The Reason, Status, and LastTransitionTime do not change.